File: ios/SiezaGym/DesignSystem/Theme.swift
import SwiftUI
import UIKit

/// Un tema del diseño: los mismos tokens que `.d2[data-d2-theme="..."]` en la
/// web. Los valores los genera `ios/scripts/sync-theme.mjs` desde el CSS, así
/// que no hay colores escritos a mano de este lado.
struct Theme: Identifiable, Equatable, Sendable {
    let id: String
    let nombre: String

    /// Opacidad de la tinta del vidrio, en tres niveles.
    let glass1: Double
    let glass2: Double
    let glass3: Double

    let borde: Color
    let bordeFuerte: Color

    let texto: Color
    let texto2: Color
    let texto3: Color

    /// El sólido del tema: el botón principal, el día entrenado, la serie
    /// confirmada. Y el color del texto que va encima.
    let solido: Color
    let sobreSolido: Color

    /// Las tres luces que se apoyan sobre el degradado de base.
    let luzA: Color
    let luzB: Color
    let luzC: Color

    let mancha1: Color
    let mancha2: Color
    let mancha3: Color

    let fondoInicio: UnitPoint
    let fondoFin: UnitPoint
    let fondo: [Gradient.Stop]

    /// Nombre de la imagen del catálogo para los temas que traen una obra de
    /// fondo en vez de un degradado. `nil` en los demás.
    let obra: String?

    static let porDefecto = Theme.todos.first { $0.id == "plata" } ?? Theme.todos[0]

    static func conId(_ id: String?) -> Theme {
        Theme.todos.first { $0.id == id } ?? .porDefecto
    }

    /// Un color plano del tema, para los pocos lugares que no pueden llevar el
    /// degradado entero (una barra de sistema, un relleno de respaldo).
    var fondoPlano: Color { fondo.last?.color ?? .black }

    /// El vidrio no tiene color propio: es blanco translúcido y el color se lo
    /// da el fondo que difumina.
    func vidrio(_ nivel: Int = 1) -> Color {
        let opacidad = nivel >= 3 ? glass3 : (nivel == 2 ? glass2 : glass1)
        return Color.white.opacity(opacidad)
    }
}

/// El tema elegido, guardado en el aparato igual que en la web.
///
/// `@AppStorage` es el equivalente de localStorage: es una preferencia del
/// teléfono, no de la cuenta, así que no viaja a Firestore.
@Observable
final class ThemeStore {
    var actual: Theme {
        didSet { UserDefaults.standard.set(actual.id, forKey: Self.clave) }
    }

    private static let clave = "d2-theme-v2"

    init() {
        actual = Theme.conId(UserDefaults.standard.string(forKey: Self.clave))
    }
}

private struct ThemeKey: EnvironmentKey {
    static let defaultValue = Theme.porDefecto
}

extension EnvironmentValues {
    var tema: Theme {
        get { self[ThemeKey.self] }
        set { self[ThemeKey.self] = newValue }
    }
}

// MARK: - Piezas

/// El fondo de la app: la obra del tema, o el degradado con sus tres luces y
/// las manchas desenfocadas encima.
///
/// Las manchas no son decoración: el vidrio de las tarjetas difumina lo que
/// tiene atrás, y sobre un color plano el desenfoque no se percibe.
struct Backdrop: View {
    @Environment(\.tema) private var tema

    var body: some View {
        GeometryReader { proxy in
            let w = proxy.size.width
            let h = proxy.size.height

            ZStack {
                LinearGradient(stops: tema.fondo, startPoint: tema.fondoInicio, endPoint: tema.fondoFin)

                if let obra = tema.obra {
                    Image(obra)
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                        .frame(width: w, height: h)
                        .clipped()
                } else {
                    // Las tres luces, en las mismas posiciones que los
                    // radial-gradient de la web.
                    luz(tema.luzA, x: 0.18, y: 0.08, radio: w * 1.1)
                    luz(tema.luzB, x: 0.88, y: 0.22, radio: w * 0.9)
                    luz(tema.luzC, x: 0.50, y: 1.05, radio: w * 1.2)

                    mancha(tema.mancha1, x: -0.14, y: 0.06, lado: 460)
                    mancha(tema.mancha2, x: 0.86, y: 0.26, lado: 380)
                    mancha(tema.mancha3, x: 0.24, y: 1.12, lado: 520)
                }
            }
            .frame(width: w, height: h)
        }
        .ignoresSafeArea()
    }

    private func luz(_ color: Color, x: Double, y: Double, radio: Double) -> some View {
        RadialGradient(colors: [color, color.opacity(0)], center: UnitPoint(x: x, y: y), startRadius: 0, endRadius: radio)
    }

    private func mancha(_ color: Color, x: Double, y: Double, lado: Double) -> some View {
        Circle()
            .fill(color)
            .frame(width: lado, height: lado)
            .blur(radius: 70)
            .position(x: x * UIScreen.main.bounds.width, y: y * UIScreen.main.bounds.height)
    }
}

/// Tarjeta de vidrio: el contenedor de todo en este diseño.
struct GlassCard<Content: View>: View {
    @Environment(\.tema) private var tema
    var padding: CGFloat = 16
    var radius: CGFloat = 24
    var nivel: Int = 1
    @ViewBuilder var content: Content

    var body: some View {
        content
            .padding(padding)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(.ultraThinMaterial.opacity(0.6), in: .rect(cornerRadius: radius))
            .background(tema.vidrio(nivel), in: .rect(cornerRadius: radius))
            .overlay {
                RoundedRectangle(cornerRadius: radius)
                    .strokeBorder(tema.borde, lineWidth: 1)
            }
    }
}

/// Rótulo de sección: el `d2-label` de la web.
struct SectionLabel: View {
    @Environment(\.tema) private var tema
    let texto: String

    init(_ texto: String) { self.texto = texto }

    var body: some View {
        Text(texto)
            .font(.system(size: 13))
            .foregroundStyle(tema.texto2)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.leading, 4)
    }
}

/// Título de pantalla.
struct PageTitle: View {
    @Environment(\.tema) private var tema
    let texto: String

    init(_ texto: String) { self.texto = texto }

    var body: some View {
        Text(texto)
            .font(.system(size: 30, weight: .heavy))
            .tracking(-0.7)
            .foregroundStyle(tema.texto)
            .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// El botón principal: el sólido del tema, con el texto que le corresponde.
struct SolidButtonStyle: ButtonStyle {
    @Environment(\.tema) private var tema
    var expands = true

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.system(size: 15, weight: .medium))
            .foregroundStyle(tema.sobreSolido)
            .padding(.horizontal, 22)
            .frame(maxWidth: expands ? .infinity : nil, minHeight: 52)
            .background(tema.solido, in: .capsule)
            .opacity(configuration.isPressed ? 0.85 : 1)
            .scaleEffect(configuration.isPressed ? 0.98 : 1)
            .animation(.snappy(duration: 0.15), value: configuration.isPressed)
    }
}

/// Botón secundario: contorno de vidrio, sin relleno sólido.
struct GhostButtonStyle: ButtonStyle {
    @Environment(\.tema) private var tema

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.system(size: 14))
            .foregroundStyle(tema.texto)
            .padding(.horizontal, 18)
            .frame(minHeight: 46)
            .background(tema.vidrio(1), in: .capsule)
            .overlay { Capsule().strokeBorder(tema.bordeFuerte, lineWidth: 1) }
            .opacity(configuration.isPressed ? 0.7 : 1)
    }
}

// MARK: - Piezas que venían del diseño anterior, ahora sobre vidrio

extension Theme {
    /// Radios del diseño, iguales a los de la web.
    static let radius: CGFloat = 24
    static let radiusSmall: CGFloat = 16
}

/// Encabezado de widget: rótulo chico arriba.
struct WidgetHeader: View {
    @Environment(\.tema) private var tema
    let title: String

    var body: some View {
        Text(title)
            .font(.system(size: 13))
            .foregroundStyle(tema.texto2)
            .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// Número grande de un widget, con su unidad al lado.
struct WidgetValue: View {
    @Environment(\.tema) private var tema
    let value: String
    var unit: String?

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: 3) {
            Text(value)
                .font(.system(size: 26, weight: .semibold))
                .tracking(-0.8)
                .foregroundStyle(tema.texto)
            if let unit {
                Text(unit)
                    .font(.system(size: 12))
                    .foregroundStyle(tema.texto2)
            }
        }
        .lineLimit(1)
        .minimumScaleFactor(0.6)
    }
}

/// Barra de progreso: riel tenue y relleno con el sólido del tema, la misma
/// que `.d2-muscle-bar` en la web.
struct WidgetMeter: View {
    @Environment(\.tema) private var tema
    /// 0...1
    let value: Double

    var body: some View {
        GeometryReader { proxy in
            let ancho = proxy.size.width
            ZStack(alignment: .leading) {
                Capsule().fill(tema.texto3.opacity(0.35))
                Capsule().fill(tema.solido).frame(width: ancho * min(max(value, 0), 1))
            }
        }
        .frame(height: 6)
    }
}

/// Alias del contenedor de tarjeta, para las pantallas que ya lo usaban.
typealias SurfaceCard = GlassCard
typealias AccentButtonStyle = SolidButtonStyle

extension View {
    /// Los teclados numéricos de iOS no traen una tecla para cerrarlos.
    /// Todas las pantallas con formularios comparten este botón para que el
    /// usuario siempre pueda terminar la edición sin buscar un gesto oculto.
    func tecladoConBotonListo() -> some View {
        toolbar {
            ToolbarItemGroup(placement: .keyboard) {
                Spacer()
                Button("Listo") {
                    UIApplication.shared.sendAction(
                        #selector(UIResponder.resignFirstResponder),
                        to: nil,
                        from: nil,
                        for: nil
                    )
                }
            }
        }
    }
}

File: android/app/src/main/kotlin/com/siezagym/app/DesignSystem/Theme.kt
package com.siezagym.app.DesignSystem

import android.content.SharedPreferences
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CapsuleShape
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.blur
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Offset
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.siezagym.app.R

/** Una parada del degradado de fondo, como `Gradient.Stop` en iOS. */
@Immutable
data class ColorStop(val color: Color, val location: Float)

/**
 * Un tema del diseño: los mismos tokens que `.d2[data-d2-theme="..."]` en la
 * web. Los valores los genera `android/scripts/sync-theme.mjs` desde el CSS,
 * así que no hay colores escritos a mano de este lado.
 */
@Immutable
data class D2Theme(
    val id: String,
    val nombre: String,

    /** Opacidad de la tinta del vidrio, en tres niveles. */
    val glass1: Float,
    val glass2: Float,
    val glass3: Float,

    val borde: Color,
    val bordeFuerte: Color,

    val texto: Color,
    val texto2: Color,
    val texto3: Color,

    /** El sólido del tema: el botón principal, el día entrenado, la serie
     * confirmada. Y el color del texto que va encima. */
    val solido: Color,
    val sobreSolido: Color,

    /** Las tres luces que se apoyan sobre el degradado de base. */
    val luzA: Color,
    val luzB: Color,
    val luzC: Color,

    val mancha1: Color,
    val mancha2: Color,
    val mancha3: Color,

    /** Puntos de inicio y fin del degradado, en fracción del alto y ancho. */
    val fondoInicio: Pair<Float, Float>,
    val fondoFin: Pair<Float, Float>,
    val fondo: List<ColorStop>,

    /**
     * Nombre del tema cuando trae una obra de fondo en vez de un degradado.
     * `null` en los demás.
     */
    val obra: String?,
) {
    /** Un color plano del tema, para los pocos lugares que no pueden llevar el
     * degradado entero (los indicadores de sistema, un relleno de respaldo). */
    val fondoPlano: Color get() = fondo.lastOrNull()?.color ?: Color.Black

    /** El vidrio no tiene color propio: es blanco translúcido y el color se lo
     * da el fondo que difumina. */
    fun vidrio(nivel: Int = 1): Color {
        val opacidad = if (nivel >= 3) glass3 else if (nivel == 2) glass2 else glass1
        return Color.White.copy(alpha = opacidad)
    }

    companion object {
        fun conId(id: String?): D2Theme =
            TemasTokens.firstOrNull { it.id == id } ?: TemaPorDefecto
    }
}

/** El tema elegido, guardado en el aparato igual que en la web. */
class ThemeStore(private val preferencias: SharedPreferences) {
    private val clave = "d2-theme-v2"

    var actual: D2Theme by mutableStateOf(D2Theme.conId(preferencias.getString(clave, null)))
        private set

    fun elegir(id: String) {
        actual = D2Theme.conId(id)
        preferencias.edit().putString(clave, id).apply()
    }

    companion object {
        val CLAVE_PREFERENCIAS = "com.siezagym.app.preferences"
    }
}

val LocalD2Theme = staticCompositionLocalOf<D2Theme> { TemaPorDefecto }
val LocalThemeStore = staticCompositionLocalOf<ThemeStore> {
    error("Sin ThemeStore en el árbol")
}

// MARK: - Piezas

/**
 * El fondo de la app: la obra del tema, o el degradado con sus tres luces y
 * las manchas desenfocadas encima.
 *
 * Las manchas no son decoración: el vidrio de las tarjetas difumina lo que
 * tiene atrás, y sobre un color plano el desenfoque no se percibe.
 */
@Composable
fun Backdrop(modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    val obra = tema.obra

    androidx.compose.foundation.layout.BoxWithConstraints(modifier = modifier.fillMaxSize()) {
        val ancho = maxWidth
        val alto = maxHeight
        val gradiente = Brush.linearGradient(
            colors = tema.fondo.map { it.color },
            colorStops = tema.fondo.map { it.location }.toFloatArray(),
            start = Offset(ancho.value * tema.fondoInicio.first, alto.value * tema.fondoInicio.second),
            end = Offset(ancho.value * tema.fondoFin.first, alto.value * tema.fondoFin.second),
        )

        Box(Modifier.fillMaxSize().background(gradiente))

        if (obra != null) {
            Obra(obra)
        } else {
            // Las tres luces, en las mismas posiciones que los
            // radial-gradient de la web.
            Luz(tema.luzA, 0.18f, 0.08f, radio = ancho * 1.1f, ancho, alto)
            Luz(tema.luzB, 0.88f, 0.22f, radio = ancho * 0.9f, ancho, alto)
            Luz(tema.luzC, 0.50f, 1.05f, radio = ancho * 1.2f, ancho, alto)

            Mancha(tema.mancha1, x = -0.14f, y = 0.06f, lado = 460.dp, ancho, alto)
            Mancha(tema.mancha2, x = 0.86f, y = 0.26f, lado = 380.dp, ancho, alto)
            Mancha(tema.mancha3, x = 0.24f, y = 1.12f, lado = 520.dp, ancho, alto)
        }
    }
}

@Composable
private fun Obra(nombre: String) {
    val pintura: Painter? = when (nombre) {
        "pliegues" -> painterResource(R.drawable.fondo_pliegues)
        "electrico" -> painterResource(R.drawable.fondo_electrico)
        else -> null
    }
    if (pintura != null) {
        androidx.compose.foundation.Image(
            painter = pintura,
            contentDescription = null,
            contentScale = ContentScale.Crop,
            modifier = Modifier.fillMaxSize(),
        )
    }
}

@Composable
private fun Luz(color: Color, x: Float, y: Float, radio: Dp, ancho: Dp, alto: Dp) {
    Box(
        Modifier
            .fillMaxSize()
            .drawBehind {
                drawRect(
                    brush = Brush.radialGradient(
                        colors = listOf(color, color.copy(alpha = 0f)),
                        center = Offset(ancho.value * x, alto.value * y),
                        radius = radio.value,
                    ),
                )
            },
    )
}

@Composable
private fun Mancha(color: Color, x: Float, y: Float, lado: Dp, ancho: Dp, alto: Dp) {
    Box(
        Modifier
            .offset(x = ancho * x, y = alto * y)
            .size(lado)
            .blur(70.dp)
            .background(color, CircleShape),
    )
}

/** Tarjeta de vidrio: el contenedor de todo en este diseño. */
@Composable
fun GlassCard(
    modifier: Modifier = Modifier,
    paddingInterno: Dp = 16.dp,
    radius: Dp = 24.dp,
    nivel: Int = 1,
    contenido: @Composable ColumnScope.() -> Unit,
) {
    val tema = LocalD2Theme.current
    Column(
        modifier = modifier
            .clip(RoundedCornerShape(radius))
            .background(tema.vidrio(nivel))
            .border(1.dp, tema.borde, RoundedCornerShape(radius))
            .padding(paddingInterno),
        horizontalAlignment = Alignment.Start,
        content = contenido,
    )
}

/** Rótulo de sección: el `d2-label` de la web. */
@Composable
fun SectionLabel(texto: String, modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    Text(
        texto,
        fontSize = 13.sp,
        color = tema.texto2,
        modifier = modifier
            .fillMaxWidth()
            .padding(start = 4.dp),
    )
}

/** Título de pantalla. */
@Composable
fun PageTitle(texto: String, modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    Text(
        texto,
        fontSize = 30.sp,
        fontWeight = FontWeight.Heavy,
        letterSpacing = (-0.7).sp,
        color = tema.texto,
        modifier = modifier.fillMaxWidth(),
    )
}

// MARK: - Botones

/** El botón principal: el sólido del tema, con el texto que le corresponde. */
@Composable
fun SolidButton(
    texto: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    expands: Boolean = true,
    enabled: Boolean = true,
) {
    val tema = LocalD2Theme.current
    Button(
        onClick = onClick,
        enabled = enabled,
        shape = CapsuleShape,
        colors = ButtonDefaults.buttonColors(
            containerColor = tema.solido,
            contentColor = tema.sobreSolido,
            disabledContainerColor = tema.solido.copy(alpha = 0.5f),
            disabledContentColor = tema.sobreSolido.copy(alpha = 0.7f),
        ),
        contentPadding = PaddingValues(horizontal = 22.dp, vertical = 0.dp),
        modifier = modifier
            .then(if (expands) Modifier.fillMaxWidth() else Modifier)
            .height(52.dp),
    ) {
        Text(texto, fontSize = 15.sp, fontWeight = FontWeight.Medium)
    }
}

/** Botón secundario: contorno de vidrio, sin relleno sólido. */
@Composable
fun GhostButton(
    texto: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val tema = LocalD2Theme.current
    Button(
        onClick = onClick,
        shape = CapsuleShape,
        colors = ButtonDefaults.buttonColors(
            containerColor = tema.vidrio(1),
            contentColor = tema.texto,
        ),
        border = BorderStroke(1.dp, tema.bordeFuerte),
        contentPadding = PaddingValues(horizontal = 18.dp, vertical = 0.dp),
        modifier = modifier.height(46.dp),
    ) {
        Text(texto, fontSize = 14.sp)
    }
}

// MARK: - Widgets

/** Encabezado de widget: rótulo chico arriba. */
@Composable
fun WidgetHeader(title: String, modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    Text(
        title,
        fontSize = 13.sp,
        color = tema.texto2,
        modifier = modifier.fillMaxWidth(),
    )
}

/** Número grande de un widget, con su unidad al lado. */
@Composable
fun WidgetValue(value: String, unit: String? = null, modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    Row(verticalAlignment = Alignment.Bottom, modifier = modifier.fillMaxWidth()) {
        Text(
            value,
            fontSize = 26.sp,
            fontWeight = FontWeight.SemiBold,
            letterSpacing = (-0.8).sp,
            color = tema.texto,
            maxLines = 1,
            softWrap = false,
        )
        if (unit != null) {
            Spacer(Modifier.width(3.dp))
            Text(unit, fontSize = 12.sp, color = tema.texto2)
        }
    }
}

/** Barra de progreso: riel tenue y relleno con el sólido del tema, la misma
 * que `.d2-muscle-bar` en la web. `proporcion` va de 0...1. */
@Composable
fun WidgetMeter(proporcion: Float, modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    Box(modifier.height(6.dp)) {
        Box(Modifier.fillMaxSize().clip(CapsuleShape).background(tema.texto3.copy(alpha = 0.35f)))
        Box(
            Modifier
                .fillMaxWidth(proporcion.coerceIn(0f, 1f))
                .fillMaxHeight()
                .clip(CapsuleShape)
                .background(tema.solido),
        )
    }
}